x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/support/SecondaryAuthenticatorTests.java

```diff
@@ -0,0 +1,295 @@
+import org.elasticsearch.xpack.core.security.action.user.AuthenticateRequest;
+import org.elasticsearch.xpack.core.security.authc.Authentication;
+import org.elasticsearch.xpack.core.security.authc.Authentication.AuthenticationType;
+import org.elasticsearch.xpack.core.security.authc.Authentication.RealmRef;
+import org.elasticsearch.xpack.core.security.authc.AuthenticationFailureHandler;
+import org.elasticsearch.xpack.core.security.authc.DefaultAuthenticationFailureHandler;
+import org.elasticsearch.xpack.core.security.authc.RealmConfig;
+import org.elasticsearch.xpack.core.security.authc.RealmConfig.RealmIdentifier;
+import org.elasticsearch.xpack.core.security.authc.support.SecondaryAuthentication;
+import org.elasticsearch.xpack.core.security.index.RestrictedIndicesNames;
+import org.elasticsearch.xpack.core.security.user.AnonymousUser;
+import org.elasticsearch.xpack.core.security.user.User;
+import org.elasticsearch.xpack.security.audit.AuditTrailService;
+import org.elasticsearch.xpack.security.authc.ApiKeyService;
+import org.elasticsearch.xpack.security.authc.AuthenticationService;
+import org.elasticsearch.xpack.security.authc.Realms;
+import org.elasticsearch.xpack.security.authc.TokenService;
+import org.elasticsearch.xpack.security.support.SecurityIndexManager;
+import org.elasticsearch.xpack.security.test.SecurityMocks;
+import org.hamcrest.Matchers;
+import org.junit.After;
+import org.junit.Before;
+import org.mockito.Mockito;
+
+import java.nio.charset.StandardCharsets;
+import java.time.Clock;
+import java.util.Base64;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+
+import static org.elasticsearch.xpack.security.authc.support.SecondaryAuthenticator.SECONDARY_AUTH_HEADER_NAME;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.nullValue;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class SecondaryAuthenticatorTests extends ESTestCase {
+
+    private AuthenticationService authenticationService;
+    private SecondaryAuthenticator authenticator;
+    private DummyUsernamePasswordRealm realm;
+    private ThreadPool threadPool;
+    private SecurityContext securityContext;
+    private TokenService tokenService;
+    private Client client;
+
+    @Before
+    public void setupMocks() throws Exception {
+        threadPool = new TestThreadPool(getTestName());
+        final ThreadContext threadContext = threadPool.getThreadContext();
+
+        final Realms realms = mock(Realms.class);
+        final Settings settings = Settings.builder()
+            .put(buildEnvSettings(Settings.EMPTY))
+            .put("xpack.security.authc.realms.dummy.test_realm.order", 1)
+            .put("xpack.security.authc.token.enabled", true)
+            .put("xpack.security.authc.api_key.enabled", true)
```

Comment:
We don't actually use ApiKeyService here so we could just mock it below and return false on `isEnabled()` ?

Reply:
Yes. I originally expected to include a test for ApiKey authentication, but Bearer Token is sufficient to verify that we're not restricted to the realm chain.

Reply:
Actually, this is not as easy as it sounds.
`authenticateWithApiKeyIfPresent` is package protected, so it's really hard to mock.